### DISCUSSION TITLE
helix vs vim rant

### DISCUSSION THREAD

**Discussion by txtyash** [helix-editor/helix#1324] [General] [answered]
I've argued a lot about keybindings having redundant amount of keystrokes for an
equivalent action in vim.

Irrespective of the fact that helix is inspired by kakoune and vim both, the set of keybindings really make sense for some actions and for others it is just an excessive amount of keystrokes with an argument that bindings should be memorable. Well, I like that idea that bindings should be memorable but in some cases we should prefer short amount of keystrokes for a binding over idiomatic ones.

As an example let's compare vim bindings and hx bindings side by side:

- `ge better than G`

- `miwc worse than ciw`: In this case I think we should swap `mi'c` with `vi'c` which sounds like visually in single quotes and change, whereas we should have `miw` doing the same action as `ciw` which might sound like modify in word. Here both `vi'c` as well a `miw` sound idiomatic and are more flexible as compared to `miwc`. Users have choice to visually select the word and then change it or a shorter alternative to change the word directly.

- `vgld` worse than `D`: This can be mapped by the user but should be availabe by default.

- `vxgegld` worse than `dG`: I know this too might be fixed. I have started a discussion previously over these minor details. But if selection action model is to be followed then it should be consistent across all movements.

- `mm` better than `%`

- `C-a` worse than `C-a`

- `C-x` worse than `C-x`

- We can do `o` and `O` better than vim:
```
    o = ["add_newline_below", "move_line_down"]
    O = ["add_newline_above", "move_line_up"]
```
I would like you(owner) to use these personally everyday and see the difference.

,etc. etc.

There are a tonne of things better done than vim but there are also some small improvements to be made. Everyone might have different opinion over those but we can definitely settle on better choices. And `you can map that manually in the config` is a really bad argument when someone complains about the keybindings.

Other areas where I think helix can be better is default functionality which it
already does better than everything else. I didn't notice any speed issues
unlike other text editors but here are some suggested improvements:
- jump/follow mode(similar to amp)
- tabout feature(similar to vscode or nvim-tabout)
* Note taking should be taken under consideration
    - orgmode for notetaking(Sorry couldn't submit an RFC)
    - Grammar checker for note taking
    - Spell checker for note taking
    - Word meanings for note taking
- yanking should have post highlighting with a timeout
- `/` search should highlight all matches & match under the cursor with a different color.
- Light themes should also be taken seriously.
- Some dark themes need minor improvements where matching braces don't get highlighted with a different color.

I am unable to use helix on a daily basis due to my exams but I check it every now and then. I cannot code for myself but I'm fairly good at using vim so I can at least think of improvements as a user.

## EDIT(Additional suggestions):
- Command section would look nicer at the top. I once made a post about it in r/neovim, some people liked the idea & [VonHeikemen](https://github.com/VonHeikemen) also made two plugins finecmdline & searchbox as a proof of concept. Everybody loved those plugins. I think it will be ergonomically better to have this built in by default in helix.

- Table mode w/o org mode: https://github.com/dhruvasagar/vim-table-mode

- When we copy/cut something and paste it at the end of the line then a space
has to be added manually unless you don't want a space. This can be fixed by
automatically adding a space between the end of the line and the content we are
pasting there and also have the cursor placed at the space so when we don't want
space we simply remove it using `d`.

- Add normal mode in file picker or buffer picker & also make it the default
mode in buffer picker. This way navigation will be easier with `j,k` and in
buffer picker it is more likely to be a faster way. 

- autopairs for `(backtick) should be added.

- I face a lot of issues with visual mode too but Ig it is due to the kakoune
mix model and I'm probably not well versed in doing things the kakoune way so
I'll add about this later.

- `xv` selects the entire line and puts in visual mode, so when I scroll down
with `j` then it keeps selecting entire lines which is expected(I know same can
be done with `xx`) but `xvk`
doesn't select the line I started with instead just selects the first character
on that line and then starts selecting entire lines upwards.

- Color preview built-in similar to [this plugin](https://github.com/norcalli/nvim-colorizer.lua). It is probably mostly going to be used with CSS files but it saves a tonne of work.

- Markdown preview similar to [this plugin](https://github.com/iamcco/markdown-preview.nvim). This is yet another lifesaver. Fast and easy.

**Comment by NNBnh** [comment]
This is a pretty well constructive criticism, but I have a question:

> * `C-a` worse than `C-a`
> * `C-x` worse than `C-x`

What? can you expand this more clearly? I'm stuck in Helix 0.5 so idk how <kbd>C-a</kbd> and <kbd>C-x</kbd> been implemented.

> * Grammar checker for note taking
> * Spell checker for note taking

This features should only be implemented for comments on code file, note taking file formats (markdown, html, latex, org mode, asciidoc...) and plain text files.

> - Word meanings for note taking

I'm not familiar with this feature so my guess is this is a personal dictionary or a dictionary lookup... Either way I afraid that this won't be built-in into Helix, maybe through plug-ins...

> * `/` search should highlight all matches & match under the cursor with a different color.
> * Light themes should also be taken seriously.
> * Some dark themes need minor improvements where matching braces don't get highlighted with a different color.

Hmm, I'll take a look into this.

**Reply by txtyash** [reply]
1. `C-a/x` in vim acts on next object, whereas in hx it acts on the object under the cursor.

2. Yes, Those grammar and spellcheck features should only be present in specific filetypes.

3. As you can see in [this picture ](https://i.imgur.com/CQP7zuH.png) how the LSP shows me the documentation for the print macro, similarly word meanings would be a nice to have with filetypes like org, markdown, etc.

**Reply by txtyash** [reply]
> Hmm, I'll take a look into this.

Hey, also try to use a different cursor shape or color for insert and normal modes. Some themes are using the same cursor color for insert as well as normal mode.

**Comment by kirawi** [comment]
I feel like the Org mode stuff should remain as a separate plugin/plugin ecosystem.

**Comment by NNBnh** [comment]
> * Table mode w/o org mode: https://github.com/dhruvasagar/vim-table-mode

A simpler approach to this would be using a code format to format table ([Prettier](https://prettier.io) support format markdown table) and later on make an option to format code in real time.

**Comment by Omnikar** [comment, marked as answer]
> `<space>p` should paste right below the cursor(always): `<space>p` is inconsistent
with where it actually pastes. If your cursor is on a non-empty line then it will
paste it the same line, whereas on a empty line it will paste on the line below
the cursor.

`<space>p` pastes after the selection, just like `p`. When you're on an empty line, pasting after the cursor means it pastes after the line break, i.e., on the next line. Pasting before the selection is `<space>P`.

**Reply by txtyash** [reply]
Oh yes. I'm not sure why it was acting weird the last time I was using it. Probably the line I was pasting had a return character in the beginning or something.

**Reply by heliostatic** [reply]
I actually think this is still inconsistent, at least to the end user? I added a comment with a video repro in an existing issue, but I'll add here as well:

https://user-images.githubusercontent.com/70898/151583102-67bd2fdd-3fee-4235-81e8-34cd88bf0dd0.mp4

**Reply by Omnikar** [reply]
I can't tell what the capitalization of the letters in that video are, at first glance it looks like they're all uppercase, but then you would have used `U` to undo…

**Comment by F0Xde** [comment]
I think spell checking is a very convenient feature to have everywhere, but of course it would need to be able to deal with naming schemes like camelCase etc.

**Reply by DrSensor** [reply]
I think spell checking should be handled by language server (e.g [ltex-ls](https://github.com/valentjn/ltex-ls)).
That's mean helix should support multiple language server (https://github.com/helix-editor/helix/issues/1396) in one instance.

**Reply by David-Else** [reply]
I really need [ltex-ls](https://github.com/valentjn/ltex-ls) to work on any Markdown files, is it possible to add it now using a config file? If so, could you explain? Thanks!

**Reply by sudormrfbin** [reply]
This file should be in the same directory as your `config.toml`:

```toml
# in <config_dir>/helix/languages.toml

[[language]]
name = "markdown"
language-server = { command = "ltex-ls" }
```

**Reply by David-Else** [reply]
Cheers! It works in that it finds errors, and I can go from error to error with `]d` great, but when I select `space a` for code actions it says there is no code action :( In Neovim it presents a list of correct spellings or corrected grammar for you to choose. I assume this needs to be fixed somehow in helix. Is it OK to make feature requests in the issues? I am afraid I know hardly any Rust and can't contribute myself at this time.

**Reply by sudormrfbin** [reply]
Of course you can open an issue :) It would also be helpful if you could run with `hx -v` and press space+a to check the log for any clues. We might be ignoring some specific kind of code action sent by the server.

**Reply by EricCrosson** [reply]
Any particular reason to keep
```toml
language-server = { command = "ltex-ls" }
```

in a personal config, or do you mind if I open a PR to make this default?

**Reply by David-Else** [reply]
@EricCrosson I would love it to be default, but check the issue I just opened: https://github.com/helix-editor/helix/issues/1942

**Reply by David-Else** [reply]
- vgld worse than D: This can be mapped by the user but should be availabe by default.

How can I do this please?

**Reply by Omnikar** [reply]
If you want to map `D` to delete to the end of the line, you'd do:
```toml
[keys.normal]
"D" = ["collapse_selection", "select_mode", "goto_line_end", "delete_selection"]
```
Note that without remapping, you can also do `t<enter>` as a replacement for `vgl`.

**Comment by juniway** [comment]
I don't think `ge` is better than `G`, because you still press `gg` to the beginning of the buffer (not `gb` to be consistent with its semantics meaning).
